registry/storage/driver/swift/swift.go

```diff
@@ -0,0 +1,519 @@
+	if err != nil {
+		if swiftErr, ok := err.(*swift.Error); ok {
+			if swiftErr.StatusCode == 404 {
+				// Create a object manifest
+				if dir, err := d.createParentFolder(path); err != nil {
+					return bytesRead, parseError(dir, err)
+				}
+				headers := make(swift.Headers)
+				headers["X-Object-Manifest"] = segmentsContainer + "/" + d.swiftPath(path)
+				manifest, err := d.Conn.ObjectCreate(d.Container, d.swiftPath(path), false, "",
+					d.getContentType(), headers)
+				manifest.Close()
+				if err != nil {
+					return bytesRead, parseError(path, err)
+				}
+			} else {
+				return bytesRead, parseError(path, err)
+			}
+		} else {
+			return bytesRead, parseError(path, err)
+		}
+	} else {
+		// The manifest already exists. Get all the segments
+		currentLength = info.Bytes
+		headers := make(swift.Headers)
+		headers["Content-Type"] = "application/json"
+		opts := &swift.ObjectsOpts{Prefix: d.swiftPath(path), Headers: headers}
+		segments, err = d.Conn.Objects(d.Container+"_segments", opts)
+		if err != nil {
+			return bytesRead, parseError(path, err)
+		}
+	}
+
+	// First, we skip the existing segments that are not modified by this call
+	for i := range segments {
+		if offset < cursor+segments[i].Bytes {
+			break
+		}
+		cursor += segments[i].Bytes
+		partNumber++
+	}
+
+	// We reached the end of the file but we haven't reached 'offset' yet
+	// Therefore we add blocks of zeros
+	if offset >= currentLength {
+		for offset-currentLength >= d.ChunkSize {
+			// Insert a block a zero
+			d.Conn.ObjectPut(segmentsContainer, getSegment(),
+				bytes.NewReader(zeroBuf), false, "",
+				d.getContentType(), nil)
+			currentLength += d.ChunkSize
+			partNumber++
+		}
+
+		cursor = currentLength
+		paddingReader = bytes.NewReader(zeroBuf)
+	} else {
+		// Offset is inside the current segment : we need to read the
+		// data from the beginning of the segment to offset
+		paddingReader, _, err = d.Conn.ObjectOpen(segmentsContainer, getSegment(), false, nil)
+		if err != nil {
+			return bytesRead, parseError(getSegment(), err)
+		}
+	}
+
+	multi := io.MultiReader(
+		io.LimitReader(paddingReader, offset-cursor),
+		io.LimitReader(reader, d.ChunkSize-(offset-cursor)),
+	)
+
+	for {
+		currentSegment, err := d.Conn.ObjectCreate(segmentsContainer, getSegment(), false, "", d.getContentType(), nil)
+		if err != nil {
+			return bytesRead, parseError(path, err)
+		}
+
+		n, err := io.Copy(currentSegment, multi)
+		if err != nil {
+			return bytesRead, parseError(path, err)
+		}
+
+		if n < d.ChunkSize {
+			// We wrote all the data
+			if cursor+n < currentLength {
+				// Copy the end of the chunk
+				headers := make(swift.Headers)
+				headers["Range"] = "bytes=" + strconv.FormatInt(cursor+n, 10) + "-" + strconv.FormatInt(cursor+d.ChunkSize, 10)
+				file, _, err := d.Conn.ObjectOpen(d.Container, d.swiftPath(path), false, headers)
+				if err != nil {
+					return bytesRead, parseError(path, err)
+				}
+				io.Copy(currentSegment, file)
```

Comment:
As far as I understand we copy data through network there, so Copy may fail with an error?